### PR TITLE
 core: dispatcher test suite and stream completion

### LINKED ISSUE #302: core: handle stream completion
Per: https://github.com/lyft/envoy-mobile/blob/68eeaf6ea04fc7d89d95b3247333a2f047f6198b/library/common/http/dispatcher.h#L80-L83

The http_dispatcher needs to keep track of closing streams and issue appropriate clean up.

### PR DESCRIPTION
**Pull request by junr03**
Description: This PR adds an initial test suite for the Dispatcher implementation. This test suite is a step in allowing us to iterate on the remaining issues (stream completion, data races, buffer passing, etc) with more confidence. Additionally this PR add logic to track stream state (local and remote) in order to track stream completion state. Once the stream is completed (both local and remote are closed) the stream is removed from the http dispatcher's stream map.

Risk Level: medium - this PR takes into account stream state to modify the http dispatchers stream map.
Testing: added initial suite of unit tests.
Fixes #302

Signed-off-by: Jose Nino jnino@lyft.com

### CODE REVIEW

File: library/common/http/dispatcher.h [hunk 1]
```diff
@@ -32,17 +32,16 @@ class Dispatcher : public Logger::Loggable<Logger::Id::http> {
...
-  envoy_status_t sendHeaders(envoy_stream_t stream, envoy_headers headers, bool end_stream);
-  envoy_status_t sendData(envoy_stream_t stream, envoy_headers headers, bool end_stream);
-  envoy_status_t sendMetadata(envoy_stream_t stream, envoy_headers headers, bool end_stream);
-  envoy_status_t sendTrailers(envoy_stream_t stream, envoy_headers headers);
-  envoy_status_t locallyCloseStream(envoy_stream_t stream);
+  envoy_status_t sendHeaders(envoy_stream_t stream_id, envoy_headers headers, bool end_stream);
+  envoy_status_t sendData(envoy_stream_t stream_id, envoy_headers headers, bool end_stream);
+  envoy_status_t sendMetadata(envoy_stream_t stream_id, envoy_headers headers, bool end_stream);
+  envoy_status_t sendTrailers(envoy_stream_t stream_id, envoy_headers headers);
+  envoy_status_t locallyCloseStream(envoy_stream_t stream_id);
```
[the remark this answers is not in the document]
Comment (goaway) [reply library/common/http/dispatcher.h]: (this is a nit)

File: library/common/http/dispatcher.h [hunk 2]
```diff
@@ -95,13 +93,19 @@ class Dispatcher : public Logger::Loggable<Logger::Id::http> {
     // An implementation option would be to have drainable header maps, or done callbacks.
     std::vector<HeaderMapPtr> metadata_;
     HeaderMapPtr trailers_;
+
+    std::atomic<bool> local_closed_{};
+    std::atomic<bool> remote_closed_{};
```
Comment (goaway) [inline library/common/http/dispatcher.h]: Shouldn't these be guaranteed to only be touched by a single thread?

File: library/common/http/dispatcher.h [hunk 3]
```diff
@@ -95,13 +93,19 @@ class Dispatcher : public Logger::Loggable<Logger::Id::http> {
...
   };
 
   using DirectStreamPtr = std::unique_ptr<DirectStream>;
 
   // Everything in the below interface must only be accessed from the event_dispatcher's thread.
   // This allows us to generally avoid synchronization.
   DirectStream* getStream(envoy_stream_t stream_id);
+  void cleanup(DirectStream& stream_id);
+  void closeLocal(envoy_stream_t stream_id, bool end_stream);
+  void closeRemote(envoy_stream_t stream_id, bool end_stream);
```
Comment (goaway) [inline library/common/http/dispatcher.h]: I think these should probably be private.

Comment (junr03) [reply library/common/http/dispatcher.h]: they are

File: library/common/http/dispatcher.h (hunk 2 above)
Comment (junr03) [reply library/common/http/dispatcher.h]: yep they are actually protected, you are right. No need to make them atomic.

File: library/common/http/dispatcher.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (junr03) [reply library/common/http/dispatcher.h]: I made all the public facing ones `stream` and the internal ones that we do not need to/should not obsfuscate stream_id.

**Comment by junr03** [comment]
Alright, this is ready for review @goaway. Note per our discussions that the implementation detail of how this PR tracks stream completion state (via local booleans right now) is subject to change once we merge https://github.com/envoyproxy/envoy/pull/7752. However, the test suite is completely agnostic of this fact.

File: .gitmodules [hunk 4]
```diff
@@ -1,3 +1,3 @@
 [submodule "envoy"]
 	path = envoy
-	url = https://github.com/envoyproxy/envoy.git
```
Comment (goaway) [inline .gitmodules]: Make sure not to merge this :)

**Comment by junr03** [comment]
This PR depends on and is currently failing because this [Envoy PR](https://github.com/envoyproxy/envoy/pull/7752) has not merged. As soon as that PR merges this PR should point to Envoy upstream master and get merged.

**Comment by junr03** [comment]
@goaway this is up-to-date. Do you mind reviewing?

File: library/common/http/dispatcher.cc [hunk 5]
```diff
@@ -9,42 +9,41 @@ namespace Http {
 Dispatcher::DirectStreamCallbacks::DirectStreamCallbacks(envoy_stream_t stream,
                                                          envoy_observer observer,
                                                          Dispatcher& http_dispatcher)
-    : stream_(stream), observer_(observer), http_dispatcher_(http_dispatcher) {}
+    : stream_id_(stream), observer_(observer), http_dispatcher_(http_dispatcher) {}
```
Comment (goaway) [inline library/common/http/dispatcher.cc]: nit: `stream_handle_`

Comment (junr03) [reply library/common/http/dispatcher.cc]: updated

File: .gitmodules (hunk 4 above)
Comment (junr03) [reply .gitmodules]: yep, updated.

**Pull request merged**